### ISSUE TITLE
Porting OWUMS verification methods 

### ISSUE THREAD

**Issue by idemarinis** [openwisp/openwisp-radius#29]
Hi everybody,
   we are planning to implement a new version in Django of the OpenWISP User Management System
   using these 3 modules: Django-freeradius, openwisp-users, openwisp-radius

   As we know, **openwisp-users** is the django app to manage operators of the web application with a profile and an Organization;
   **openwisp-radius** is the django app to manage multi organizations for radius database and 
   **Django-freeradius** is the django app to manage directly freeradius database. 

   openwisp-users should implement user registration and verification via multiple methods such as _mobile phone call_, _credit card_ check and _saml2_ assertions. 

   openwisp-users should also implements global and per openwisp-user stats 
 
   Now we want to start to implement the phone call check, so we are asking 
   guidelines to integrate our code.

Thank you in advance
Regards

Ilaria

**Comment by nemesisdesign** [comment]
Hey @idemarinis,

sorry for my delayed reply.

Check out the documentation about how to enable the registration feature in django-freeradius/openwisp-radius:
https://django-freeradius.readthedocs.io/en/latest/general/registration.html

openwisp-users doesn't know anything about registering new users because it only deals with supporting multi-tenancy, for this reason is better to keep the registration logic in openwisp-radius, since this is the only module right now that allows public users to create an account which allows them to use a part of the features of OpenWISP 2. A [registration view using and extending django-rest-auth is already present in openwisp-radius](https://github.com/openwisp/openwisp-radius/blob/master/openwisp_radius/api/views.py#L114-L125).

That said, we still need to add a field to store the phone number of each user.
We could add this field in openwisp-users or openwisp-radius, openwisp-users will be less clean but still acceptable and easier.
In my opinion the field should be generic and not care about which international prefixes are supported because that should be the job of the registration API which should be configurable so users/developers can configure which international prefixes or phone number format they support.
The registration API will validate, accept or rejects phone numbers according to its configuration.

Once that the phone number field is added, some additional steps are needed:

- the registration API should be tweaked so that new users are created with `is_active` set to `False`, maybe there's a way to do that with a configuration of django-rest-auth
- when this feature is enabled, the registration API has to allow (and require) the phone number to be sent during the POST (by default django-rest-auth only checks username, email, password1 and password2)
- we need to implement a background python worker which re-implements the SIP machine logic of OWUMS and flags the user as `is_active` when it receives a phone call from a phone number of an unverified user
- we need to add a websocket endpoint where javascript logic can connect to to check the status of the verification (OWUMS implemented that with long-polling, but that doesn't scale well), this websocket endpoint should be initialized only if this feature is enabled
- we need to implement a management command which can be put in a cron (or a celery-beat configuration) which removes unverified users

Then, regarding the UI, openwisp-radius doesn't have a user interface that allows users to register and I think that's good, because it can be implemented directly in the captive page with javascript calls (additional CORS configuration is needed in django or on nginx to support this); for example, if you are still using the [ansible-freeitaliawifi-login-page](https://github.com/openwisp/ansible-freeitaliawifi-login-page) role, it could be implemented there.

I've implemented the registration directly in the login page recently and it's working pretty well.
I have also been thinking about creating a ReactJS based version of that login page role but I don't have time to do that now, maybe it could come at some point in the future.

Regarding statistics, I suggest to think and discuss this later as it needs to be implemented differently than how it was implemented in OWUMS, the old implementation is too slow to be useful to organizations which have many users.
